### PR TITLE
Add processor extended state

### LINKED ISSUE #9: Support Processor Extended State Enumeration Sub-leaf (EAX= 0DX, ECX = 1)
From the [Intel reference](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf):

#  EAX 

- Bit 00: XSAVEOPT is available.
- Bit 01: Supports XSAVEC and the compacted form of XRSTOR if set.
- Bit 02: Supports XGETBV with ECX = 1 if set.
- Bit 03: Supports XSAVES/XRSTORS and IA32_XSS if set.
- Bits 31 - 04: Reserved.

# EBX
 
- Bits 31 - 00: The size in bytes of the XSAVE area containing all states enabled by XCRO | IA32_XSS.

# ECX 

- Bits 31 - 00: Reports the supported bits of the lower 32 bits of the IA32_XSS MSR. IA32_XSS[n] can be set to 1 only if ECX[n] is 1.
- Bits 07 - 00: Used for XCR0.
- Bit 08: PT state.
- Bit 09: Used for XCR0.
- Bits 31 - 10: Reserved.

# EDX 

- Bits 31 - 00: Reports the supported bits of the upper 32 bits of the IA32_XSS MSR. IA32_XSS[n+32] can be set to 1 only if EDX[n] is 1.
- Bits 31 - 00: Reserved.

### PR DESCRIPTION
**Pull request by shepmaster**
Closes #9

### CODE REVIEW

**Comment by shepmaster** [comment]
@gnzlbg if you have time, I'd appreciate you giving this PR a once-over. I'm not sure how you might want to use this particular leaf.

File: src/lib.rs [hunk 1]
```diff
@@ -640,6 +645,113 @@ impl fmt::Debug for StructuredExtendedInformation {
...
+        })
+    }
+}
+
+#[derive(Copy,Clone)]
+pub struct ProcessorExtendedStateSecondary {
+    eax: u32,
+    ebx: u32,
+    ecx: u32,
+}
```
Comment (gnzlbg) [inline src/lib.rs:713]: `edx` is missing here: 

> Bits 31 - 00: Reports the supported bits of the upper 32 bits of the IA32_XSS MSR. IA32_XSS[n+32] can be set to 1 only if EDX[n] is 1.

---

Should also add a function returning a `u32` below to access this.

File: src/lib.rs [hunk 2]
```diff
@@ -640,6 +645,113 @@ impl fmt::Debug for StructuredExtendedInformation {
...
+
+impl ProcessorExtendedStateSecondary {
+    fn new() -> ProcessorExtendedStateSecondary {
+        let (a, b, c, _) = cpuid_ext(RequestType::ProcessorExtendedState, 0x00000001);
+        ProcessorExtendedStateSecondary { eax: a, ebx: b, ecx: c }
+    }
+
+    bit!(eax, {
+        0 => xsaveopt_available,
```
Comment (gnzlbg) [inline src/lib.rs]: I'd just call these ones after their `CPUID` names in the Intel reference:

0: `xsaveopt`
1: `xsavec` (implies `xrstor` support for compact states; there is no `xrstorc` instruction)
2: `xgetbv`
3: `xsaves` (implies `xrstors`)

**Review by gnzlbg — COMMENTED** [review]
I've just noticed some minor naming nitpicks but otherwise LGTM. 

There are states beyond the secondary but I have never needed them. It doesn't seem that one could do anything more useful than just dump the registers to the user though. At that point they might just call the `cpuid` intrinsic directly.

File: src/lib.rs (hunk 1 above)
Comment (shepmaster) [reply src/lib.rs:713]: Walk me through the usefulness of this. The next line of the manual says:

> Bits 31 - 00: Reserved.

I'm not immediately seeing the value in returning 32bits of reserved values.

Comment (gnzlbg) [reply src/lib.rs:713]: > The next line of the manual says:

I did not read that line 😆

File: src/lib.rs (hunk 2 above)
Comment (shepmaster) [reply src/lib.rs]: I'm going to change these a little (not sure why I put `available`, that's kind of the point of CPUID), but I do like to have a wee bit more verbosity.

**Pull request merged**